File: app/src/main/java/com/ptithcm/thuan6420/basecleanarchitecture/ui/login/LoginFragment.java
package com.ptithcm.thuan6420.basecleanarchitecture.ui.login;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.navigation.fragment.NavHostFragment;

import com.ptithcm.thuan6420.basecleanarchitecture.R;
import com.ptithcm.thuan6420.basecleanarchitecture.databinding.FragmentLoginBinding;

public class LoginFragment extends Fragment implements View.OnClickListener {

    private FragmentLoginBinding binding;
    private UserController userController;

    private void initController() {
        userController = new UserController(this);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater
            , ViewGroup container
            , Bundle savedInstanceState) {

        binding = FragmentLoginBinding.inflate(inflater, container, false);
        initController();
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        binding.tvToRegister.setOnClickListener(this);
        binding.btnLogin.setOnClickListener(this);
        binding.layoutLoginFragment.setOnClickListener(this);
    }

    @Override
    public void onClick(View v) {
        userController.handleOnClick(v);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();

        binding = null;
    }

    public void getValue() {
        userController.setUserEmail(binding.etEmailLogin.getText().toString().trim());
        userController.setUserPassword(binding.etPasswordLogin.getText().toString().trim());
    }

    public void showErrorEmail(String errorMessage) {
        binding.etEmailLoginLayout.setError(errorMessage);
    }

    public void showErrorPassword(String errorMessage) {
        binding.etPasswordLoginLayout.setError(errorMessage);
    }

    public void navigateToMainActivity() {
        NavHostFragment.findNavController(LoginFragment.this)
                .navigate(R.id.action_loginFragment_to_mainActivity);
    }

    public void navigateToRegisterFragment() {
        NavHostFragment.findNavController(LoginFragment.this)
                .navigate(R.id.action_loginFragment_to_registerFragment);
    }

    public void turnOnLoading() {
        binding.progressBarLogin.setVisibility(View.VISIBLE);
    }

    public void turnOffLoading() {
        binding.progressBarLogin.setVisibility(View.GONE);
    }

    public void setEnabledButton(boolean enabled) {
        binding.btnLogin.setEnabled(enabled);
    }
}

File: app/src/main/java/com/ptithcm/thuan6420/basecleanarchitecture/ui/login/LoginFragment.kt
package com.ptithcm.thuan6420.basecleanarchitecture.ui.login

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.navigation.fragment.NavHostFragment
import com.ptithcm.thuan6420.basecleanarchitecture.R
import com.ptithcm.thuan6420.basecleanarchitecture.ui.components.UiController
import com.ptithcm.thuan6420.basecleanarchitecture.ui.dialogs.DialogController
import com.ptithcm.thuan6420.basecleanarchitecture.ui.dialogs.IClickOnButtonDialogListener

class LoginFragment : Fragment(), LoginListener, IClickOnButtonDialogListener {
    private lateinit var mLoginView: LoginView

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        mLoginView = LoginViewImp(inflater, container)
        return mLoginView.getRootView()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        mLoginView.setListener(this)
    }

    private fun navigateToMainActivity() {
        NavHostFragment.findNavController(this@LoginFragment)
            .navigate(R.id.action_loginFragment_to_mainActivity)
    }

    private fun navigateToRegisterFragment() {
        NavHostFragment.findNavController(this@LoginFragment)
            .navigate(R.id.action_loginFragment_to_registerFragment)
    }

    override fun onClickLayout() {
        UiController().closeKeyBoard(this.requireActivity())
    }

    override fun onClickButtonLogin(
        invalidFields: List<InvalidField>?,
        email: String,
        password: String
    ) {
        var countErrorField: Byte = 0
        for (invalidField in invalidFields!!) {
            if (invalidField.message != null) {
                countErrorField++
            }
        }
        if (countErrorField == UserController().NO_FIELD) {
            val responseUI = UserController().isLoginSuccess(email, password)

            if (responseUI.success) {
                DialogController().showSuccessDialog(this.context, this, responseUI.message)
            } else {
                DialogController().showErrorDialog(this.context, this, responseUI.message)
            }
        }
    }

    override fun onClickTvToRegister() {
        navigateToRegisterFragment()
    }

    override fun findInvalidFields(email: String, password: String) =
        UserController().findInvalidLoginFields(email, password)

    override fun clickOnButtonDialog(id: Int) {
        if (id == R.id.btnSuccess) {
            navigateToMainActivity()
        }
    }
}